nixos/modules/security/pam.nix

```
@@ -1068,6 +1082,57 @@ in
       };
     };
 
+    security.pam.rssh = {
+      enable = lib.mkEnableOption ''
+        authenticating using a signature performed by the ssh-agent.
+      '';
+
+      settings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType = moduleSettingsType;
+          options = {
+            auth_key_file = lib.mkOption {
+              type = with lib.types; nullOr nonEmptyStr;
+              description = ''
+                Path to file with trusted public keys in OpenSSH's `authorized_keys` format. The following
+                variables are expanded to the respective PAM items:
+
+                - `service`: `PAM_SERVICE`, the service name,
+                - `user`: `PAM_USER`, the username of the entity under whose identity service will be given,
+                - `tty`: `PAM_TTY`, the terminal name,
+                - `rhost`: `PAM_RHOST`, the requesting hostname, and
+                - `ruser`: `PAM_RUSER`, the requesting entity.
+
+                These PAM items are explained in {manpage}`pam_get_item(3)`.
+
+                Variables may be specified as `$var`, `''${var}` or `''${var:defaultValue}`.
+
+                ::: {.note}
+                Specifying user-writeable files here results in an insecure configuration: a malicious process
+                can then edit such an `authorized_keys` file and bypass the ssh-agent-based authentication.
+
+                This option is ignored if {option}`security.pam.rssh.settings.authorized_keys_command` is set.
+
+                If both this option and {option}`security.pam.rssh.settings.authorized_keys_command` are unset,
+                the keys will be read from `''${HOME}/.ssh/authorized_keys`, which should be considered
+                insecure.
+              '';
+              default = "/etc/ssh/authorized_keys.d/$ruser";
```

nit: What do you think about this?

```suggestion
              default = "/etc/ssh/authorized_keys.d/\${ruser:nobody}";
```

This is how I have `rssh` currently configured, but I don't remember exactly why. Maybe there was some situation where `ruser` was unset, and that caused an error? I also think it's acceptable as-is.

---

I just fear that if it is unset, it would fall back to reading the authorised keys for user `nobody`, which might be unexpected and could open a security loop hole in certain unusual configurations. If we really need a fallback, I would prefer `root` as somewhat less dangerous – an attacker already in possession of keys permitting remote log in as root can wreak havoc even without relying on pam_rssh.

Otherwise I would prefer an error, if `$ruser` cannot be expanded, over falling back to a path which is not guaranteed to contain no keys.